Add feature to add contacts and contact names to the address book

laugengebaeck opened the pull request:
This closes #385 and #386.

Commit:
Merge GitHub Action fixes into the master-branch (#378)

* Equalize branches

* Revert "Equalize branches "

This reverts commit 0c92e9b5963dd32c75041f4d10ae23473bee2888.

* Revert "Merge branch 'master' into develop"

This reverts commit 322ba3bf8f673d579e51c0f7b45405c6080a1c9f.

* Fix CI

Co-authored-by: Julia Joch<julia.joch@student.hpi.de>
Co-authored-by: Veladus <niko.hastrich@student.hpi.de>
Co-authored-by: MrHenrik <henrik.guhl@student.hpi.de>

sm1lla committed:
Added feature for adding contacts to address book

elenagensch commented on packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/buildButtonAddContact.with.with.with..st:
```diff
@@ -0,0 +1,8 @@
+toolbuilder
+buildButtonAddContact: aFrame with: aLabel with: anAction with: builder
```
rename buildButtonAddContact to buildButtonManipulateContacts or buildButtonModifyContacts

elenagensch commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/keyEvent..st:
```diff
@@ -0,0 +1,11 @@
+keyboard handling
+keyEvent: aKey
+	
+	| shouldLogin |
```
shouldAccept

elenagensch commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/keyEvent..st:
```diff
@@ -0,0 +1,11 @@
+	| shouldLogin |
+	
+	Smalltalk os platformName = 'Win32'
+		ifTrue: [shouldLogin := (aKey keyValue = 13)]
```
shouldAccept

elenagensch commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/keyEvent..st:
```diff
@@ -0,0 +1,11 @@
+	
+	Smalltalk os platformName = 'Win32'
+		ifTrue: [shouldLogin := (aKey keyValue = 13)]
+		ifFalse: [shouldLogin := (aKey keyValue = 36)].
```
shouldAccept

elenagensch commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/keyEvent..st:
```diff
@@ -0,0 +1,11 @@
+		ifTrue: [shouldLogin := (aKey keyValue = 13)]
+		ifFalse: [shouldLogin := (aKey keyValue = 36)].
+
+	shouldLogin
```
shouldAccept

elenagensch commented on packages/IMAPClient-Core.package/ICAddressBook.class/instance/addEmail..st:
```diff
@@ -1,5 +1,4 @@
 
-	self addIfNotPresent: aString
-	
+	self addAllFirstUnlessAlreadyPresent: {aString} asOrderedCollection
```
Why hasn't it been an ordered collection before?

sm1lla replied:
We wanted to add the new element to the front of the string and  check if it is already present at the same time. This was the only function in OrderedCollection that implements it and it takes an OrderedCollection

elenagensch commented on packages/IMAPClient-Core.package/ICAddressBook.class/instance/initialize.st:
```diff
@@ -0,0 +1,3 @@
+initialize
+initialize
+	
```
newline part of our coding style? not consistent

elenagensch commented on packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/buildAddressBookPanel..st:
```diff
@@ -0,0 +1,9 @@
+		model: self;
+		frame: (0.0 @ 0.0 extent: 1 @ self buttonOffset);
+		children: {
+			self buildButtonAddContact: (0.0 @ 0.0 extent: 1 @ 1) with: 'Add Contact' with: #addContact with: builder.}
```
usage of buttonOffset possible, 0.0 to 0 or is there a reason for float?

elenagensch commented on packages/IMAPClient-UI.package/ICContactDialog.class/class/open..st:
```diff
@@ -0,0 +1,8 @@
+as yet unclassified
```
as yet unclassified

elenagensch commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/email.st:
```diff
@@ -0,0 +1,3 @@
+accessing
+email
+	^ email
```
newline between email and ^email

elenagensch reviewed (commented):
The code looks good. Well done! 👏

elenagensch committed:
Contact Objects have been introduced for address book entries. A contact consists of a name and an email address field. 

Contact names are displayed (additionally to email addresses) in the address book.

Tests for contacts and the address book are created/modified accordingly. 

This commit includes refactoring for adding contacts.

elenagensch commented on packages/IMAPClient-Core.package/ICEmail.class/instance/senderName.st:
```diff
@@ -0,0 +1,6 @@
+
+	| fromCollection |
+	fromCollection := self from findBetweenSubStrs: '<>'.
+	^ (fromCollection at: 1 ifAbsent: [ '' ])
```
from is a string that begins with the sender's name, followed by the email in `<>`"Der Tag - Der SPIEGEL <der-tag@newsletter.spiegel.de>". findBetweenSubstrs splits from in two entries. If the name is absent, senderName returns the email address.

elenagensch replied:
`<>` is s bit confusing, maybe we should introduce a constant (something like emailAdddressDelimeters) that returns `'<>'`

sm1lla replied:
I agree

elenagensch commented on packages/IMAPClient-Core.package/ICAddressBook.class/instance/readChildFoldersMailsFrom..st:
```diff
@@ -1,6 +1,8 @@
-				self addEmail: childFolderEmail senderAddressFromHeader]]
\ No newline at end of file
+				anICContact := ICContact newWith: childFolderEmail senderAddressFromHeader with: childFolderEmail senderName.
+				self addEmail: anICContact asString]]
```
contacts are passed as Strings

elenagensch commented on packages/IMAPClient-Core.package/ICContact.class/class/newWith.with..st:
```diff
@@ -0,0 +1,8 @@
+		email: anEmail;
+		name: aName;
+		yourself
+		
```
NewLine needs to be removed

elenagensch commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/emailContact..st:
```diff
@@ -0,0 +1,3 @@
+accessing
+emailContact: anObject
+	emailContact := anObject
```
newline between method name and body

elenagensch commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/emailContact.st:
```diff
@@ -0,0 +1,3 @@
+accessing
+emailContact
```
newline between method name and body

elenagensch commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/setUp.st:
```diff
@@ -2,4 +2,6 @@ running
\ No newline at end of file
+	self emailContact: 
+		(ICContact newWith: 'erika.musterfrau@rocks.de' with: 'Erika Musterfrau') asString
+	
```
remove newline

elenagensch commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/testAddEntry.st:
```diff
@@ -1,7 +1,7 @@
+	self addressBook addEmail: self emailContact.
 	
 	self assert: self addressBook size equals: 1
 	
```
remove newline

elenagensch commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/testAddEntryDuplicated.st:
```diff
@@ -1,8 +1,8 @@
+	self addressBook addEmail: self emailContact.
 	
 	self assert: self addressBook size equals: 1
 	
```
remove newline

elenagensch commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/testFindEntry.st:
```diff
@@ -1,7 +1,7 @@
 	
-	self assert: (self addressBook find: self emailAddress) equals: self emailAddress
+	self assert: (self addressBook find: self emailContact) equals: self emailContact
 	
```
remove newline

elenagensch commented on packages/IMAPClient-Tests.package/ICContactTest.class/instance/setUp.st:
```diff
@@ -0,0 +1,7 @@
+	self contact: (ICContact new 
+					email: 'erika.musterfrau@rocks.de';
+					name: 'Erika Musterfrau';
+					yourself)
```
At first I wanted to test also `newWith: with:` (class method of ICContact). Should we add this test? Otherwise we should instantiate the test object with `newWith: with:`

sm1lla replied:
I think it's fine if we just use `newWith: with:` here

elenagensch replied:
Ok

sm1lla commented on packages/IMAPClient-Core.package/ICAddressBook.class/instance/feedMeWithEntries..st:
```diff
@@ -1,6 +1,6 @@
 feedMeWithEntries: anEmailCollection
-
+	
 	anEmailCollection do: [:emailAccount | 
```
whitespace before and after square brackets

sm1lla commented on packages/IMAPClient-Core.package/ICAddressBook.class/instance/readRootFoldersMailsFrom..st:
```diff
@@ -1,5 +1,8 @@
-readRootFoldersMailsFrom: anEmailAccount
+readRootFoldersMailsFrom: anEmailAccount 
 
-	anEmailAccount rootFolder emails do: [:rootFolderEmail | 
```
whitespace before and after square brackets

sm1lla commented on packages/IMAPClient-Core.package/ICContact.class/class/newWith.with..st:
```diff
@@ -0,0 +1,8 @@
+newWith: anEmail with: aName
+	
+	^ (ICContact new)
+		email: anEmail;
```
new line after receiver of cascade

sm1lla commented on packages/IMAPClient-Core.package/ICContact.class/properties.json:
```diff
@@ -0,0 +1,15 @@
+	"pools" : [
+		 ],
+	"super" : "Object",
+	"type" : "normal" }
```
Does ICContact already have a class comment? If not, please add.

elenagensch replied:
Ok

sm1lla commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/testDeleteNonExistentEntryRaisesError.st:
```diff
@@ -1,4 +1,4 @@
 
-	self should: [self addressBook remove: self emailAddress] raise: NotFound
\ No newline at end of file
+	self should: [self addressBook remove: self emailContact] raise: NotFound
```
whitespace before and after square brackets

sm1lla commented on packages/IMAPClient-Tests.package/ICAddressBookTest.class/instance/testFindNonExistentEntryRaisesError.st:
```diff
@@ -1,4 +1,4 @@
 
-	self should: [self addressBook find: self emailAddress] raise: Error
\ No newline at end of file
+	self should: [self addressBook find: self emailContact] raise: Error
```
whitespace before and after square brackets

blackjack2693 replied:
resolved by the style guide

sm1lla commented on packages/IMAPClient-Tests.package/ICContactTest.class/instance/setUp.st:
```diff
@@ -0,0 +1,7 @@
+setUp
+	
+	self contact: (ICContact new 
+					email: 'erika.musterfrau@rocks.de';
```
new line after reciever of cascade

blackjack2693 replied:
resolved by the style guide

sm1lla commented on packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/buildAddressBookEntryList..st:
```diff
@@ -14,4 +14,6 @@ buildAddressBookEntryList: builder
\ No newline at end of file
+					item preferredWidthOfColumn: 1]) max].
+				[:listMorph | (listMorph filteredItems collect: [:item | 
+					item preferredWidthOfColumn: 2]) max].}
```
whitespace before and after square brackets

sm1lla commented on packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/buildAddressBookPanel..st:
```diff
@@ -6,4 +6,4 @@ buildAddressBookPanel: builder
 		children: {
-			self buildButtonAddContact: (0.0 @ 0.0 extent: 1 @ 1) with: 'Add Contact' with: #addContact with: builder.}
\ No newline at end of file
+			self buildButton: (0 @ 0 extent: 1 @ 1) with: 'Add Contact' with: #addContact with: builder.}
```
whitespace before and after curly brackets

sm1lla commented on packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/buildButton.with.with.with..st:
```diff
@@ -1,5 +1,5 @@
+buildButton: aFrame with: aLabel with: anAction with: builder
 	
 	^ builder pluggableButtonSpec new
 		model: self;
```
new line after receiver of cascade

blackjack2693 replied:
resolved by the updated style guide

sm1lla commented on packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/buildWith..st:
```diff
@@ -5,6 +5,6 @@ buildWith: builder
-			extent: 350 @ 200;
+			extent: 550 @ 400;
 			children: {self buildAddressBookEntryList: builder. 
 						self buildAddressBookPanel: builder})
```
whitespace before and after curly brackets

blackjack2693 replied:
correction of style guide should have resolved this

sm1lla commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/buildEmail..st:
```diff
@@ -0,0 +1,11 @@
+toolbuilder
+buildEmail: builder
+
+	^ builder pluggableInputFieldSpec new
```
new line after receiver of cascade

blackjack2693 replied:
resolved by the updated style guide

sm1lla commented on packages/IMAPClient-UI.package/ICContactDialog.class/instance/keyEvent..st:
```diff
@@ -1,11 +1,11 @@
+		ifFalse: [shouldAccept := (aKey keyValue = 36)].
 
-	shouldLogin
+	shouldAccept
```
whitespace before and after square brackets

blackjack2693 replied:
resolved by the updated style guide

sm1lla commented:
We should all memorize the style guides. I think we forgot to pay attention to some aspects in the last reviews.

blackjack2693 reviewed (approved):
1. style guide rules should be fixed, but that is not critical, just makes the code more beautiful, the rest looks good.👍

elenagensch committed:
Refactoring for contact names and add-contact feature

laugengebaeck commented:
Since nearly all requested changes have been resolved, I would merge this tomorrow :)

sm1lla reviewed (approved):
Looks good.
